### PR TITLE
feat(request): Introduce `matchedRoutes` and `routePath` for `HonoRequest`

### PR DESCRIPTION
**Pull request by usualoma**
```ts
app.use('*', async function logger(c, next) {
  await next()
  c.req.matchedRoutes.forEach(({ handler, method, path }, i) => {
    const name = handler.name || (handler.length < 2 ? '[handler]' : '[middleware]')
    console.log(
      method,
      ' ',
      path,
      ' '.repeat(Math.max(10 - path.length, 0)),
      name,
      i === c.req.routeIndex ? '<- respond from here' : ''
    )
  })
})
app.use('*', async function namedMiddleware(c, next) {
  await next()
})
app.get('/*', async (c, next) => await next())
app.get('/hi/:name', (c) => c.json({ name: c.req.param('name'), path: c.req.routePath }))
```

```
% bun run --hot app.ts
Started server http://localhost:3000
ALL   /*          logger 
ALL   /*          namedMiddleware 
GET   /hi/*       [middleware] 
GET   /hi/:name   [handler] <- respond from here
```

```
% curl http://localhost:3000/hi/hono
{"name":"hono","path":"/hi/:name"}
```

### Author should do the followings, if applicable

- [x] Add tests
- [x] Run tests
- [x] `yarn denoify` to generate files for Deno

### CODE REVIEW

**Comment by yusukebe** [comment]
@usualoma 

Could you resolve the conflicts?

**Comment by usualoma** [comment]
@yusukebe Oops. Resolved!

File: src/request.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
 /* eslint-disable @typescript-eslint/no-non-null-assertion */
 import type { ParamIndexMap, ParamStash, Params } from './router'
```
Comment (yusukebe) [inline src/request.ts:3]: Can you remove these lines?

```ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { ParamIndexMap, ParamStash, Params } from './router'
```

**Review by yusukebe — CHANGES_REQUESTED** [review]

**Comment by yusukebe** [comment]
Hi @usualoma 

This is great. I like the minimal changes to the code and the features are implemented, `Context` and routers are not changed.

I made only one comment. Check it. Thanks!

**Comment by usualoma** [comment]
@yusukebe Thank you for your review. These changes would improve DX.
Fixed problem in 4c596ac!

**Review by yusukebe — APPROVED** [review]

**Comment by yusukebe** [comment]
Thanks!

Okay, merge now. Maybe I'll release the next minor version "v3.11.0" soon.

**Pull request merged**